pkg/didcomm/protocol/didexchange/service.go

```diff
@@ -63,21 +112,115 @@ func (s *Service) Handle(msg dispatcher.DIDCommMsg) error {
 	if !current.CanTransitionTo(next) {
 		return fmt.Errorf("invalid state transition: %s -> %s", current.Name(), next.Name())
 	}
-	// TODO: call pre-transition listeners -  Issue: https://github.com/hyperledger/aries-framework-go/issues/140
-	followup, err := next.Execute(msg)
+
+	eventsTriggered := false
+	if !msg.Outbound {
+		// trigger the actionEvent if the message type is registered (if it's not outbound)
+		eventsTriggered, err = s.sendEvent(msg, thid, next)
+		if err != nil {
+			return fmt.Errorf("send events failed: %w", err)
+		}
+	}
+
+	// if no events are trigger continue the execution
+	if !eventsTriggered {
+		return s.handle(&message{Msg: msg, ThreadID: thid, NextStateName: next.Name()})
+	}
+
+	return nil
+}
+
+// RegisterActionEvent on DID Exchange protocol messages. The events are triggered for incoming ConnectionRequest,
+// ConnectionResponse or ConnectionAck message types. The consumer need to invoke the callback to resume processing.
+// Only one channel can be registered for the action events. If called multiple times, the events will be sent to the
+// last channel. This works in conjunction with RegisterAutoExecute(). If this function is called after
+// RegisterAutoExecute(), the service will not execute the processing automatically and it'll wait for the callback. If
+// both are not set, then service will not handle the messages and throw an error.
+func (s *Service) RegisterActionEvent(ch chan<- event.DIDCommEvent) error {
```

Comment:
Same here w.r.t. naming - suggest "RegisterForEventsStream" (the other one should stay with "RegisterForMsgStream")

Reply:
I would drop "For"

Reply:
Discussion is tracker [here](https://github.com/hyperledger/aries-framework-go/pull/240#discussion_r325329926).

Reply:
@troyronda the intent is to register a channel via which events would be received by the caller of this function.

`RegisterEventStream(chan event.DIDCommEvent)` reads as though whoever is calling this function is going to produce the events.

`RegisterForEventStream(chan event.DIDCommEvent)` removes that ambiguity.

Also, "Stream" is probably redundant. Instead, just say "Events" - the "stream" is implicit.

Just my 2c.